sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/adapter/ModelFieldsAdapters.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.sdk.nexmark.model.sql.adapter;
+
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.beam.sdk.nexmark.model.Auction;
+import org.apache.beam.sdk.nexmark.model.Bid;
+import org.apache.beam.sdk.nexmark.model.Person;
+
+/**
+ * Maps Java model classes to Beam SQL record types.
+ */
+public class ModelFieldsAdapters {
+
+  public static final Map<Class, ModelFieldsAdapter> ADAPTERS =
+      ImmutableMap.<Class, ModelFieldsAdapter>builder()
+          .put(Auction.class, auctionAdapter())
+          .put(Bid.class, bidAdapter())
+          .put(Person.class, personAdapter())
+          .build();
+
+  private static ModelFieldsAdapter<Person> personAdapter() {
+    return new ModelFieldsAdapter<Person>(
+        BeamRecordSqlTypeBuilder.of()
+            .withLongField("id")
+            .withStringField("name")
```

Comment:
<!--thread_id:cc_151126910_t; commit:050bdab2afe544b8a5c58bfe1dfe1914a9820c8f; resolved:0-->
<!--section:context-quote-->
> **echauchot** wrote:
> Related to my last comment: having the adapter logic in the `Person` class would avoid spreading the knowledge of the `Person` fields outside the `Person` class.

<!--section:body-->
I considered it but for the reasons in my other comment I decided against it.

This knowledge is essential for the SQL queries though. Queries need to know how to get data from records. These are just utility classes which determine this field mapping.

Reply:
Separating converter from pojo is a good thing for coupling as we discussed. But if pojo change we will have to change the code in plenty other places, it's less good for encapsulation. There is still a way to keep the field knowledge inside the pojo and keep the converter outside the pojo. Have you considered using reflexion for that? But I don't know it it's worth it. WDYT?

Reply:
I am not sure I understand how we can keep the field knowledge inside the Pojos but conversion logic outside, could you share an example code snippet?

How I see it:

Main problem is that to construct a BeamRecord you need the Pojo field values in a speciofic order, matching the order of the fields in the record type. Record type (and the order of its fields) is defined by what specific SQL query expects as an input. This means that Pojo fields retrieval has to be done at the moment of the actual conversion, when we have both the Pojo instance and the record type. 

So now the question is reduced to how to get the Pojo field values in the correct order. It can probably be done with something like this at conversion time (pseudocode):

```java
public BeamRecord pojoToRecord(T pojo) {
  BeamRecordSqlType recordType = getRecordTypeForPojo(pojo);

  List<Object> fieldValues = getFieldNames(recordType)
       .map(fieldName -> getFieldValue(pojo))
       .collect(toList());

  return new BeamRecord(recordType, fieldValues);
}
```

Now the question becomes how you do `getFieldValue(pojo, fieldName)`. 
Doing it with reflection at runtime or code generation is one way. This is probably the right way but it feels like it's out of scope here.

Another way would be to define a mapping somewhere, like this (pseudocode):
```java
class Person {
  Map<String, Supplier<Object> FIELD_VALUES_SUPPLIERS = ImmutableMap.of(
     "name", Person::getName,
     "address", Person::getAddress);

  public String getName() { ... }
  public String getAddress() { ... }
```
Something along these lines could be generic enough to keep in Pojo. Problem with this is I don't know how to express this in a readable way in Java7.

Instead of the above approaches I decided to declare the record types and Pojo field retrieval logic in one place, so that at least:
 - it's easy to see how specific Pojo fields map to specific record fields;
 - it's relatively easy to see and debug if there's a mismatch between types or field names;
 - it's easy to argue about performance of this piece;
 - doesn't require a lot of complicated code in the models or infrastructure;

Reply:
You're right it is out of the scope of this PR. all my point was to make the fields mapping generic enough to be included in sql package (like you extracted the other feature) rather than coding it for each pojo. Like you said above implementing `getFieldValue(pojo, fieldName)` using reflection might be a good thing if it is feasible.